p2p: When close to the tip, download blocks in parallel from additional peers to prevent stalling

mzumsande opened the pull request:
**Problem:**
For stalling at the tip, we have a parallel download mechanism for compact blocks that was added in #27626.
For stalling during IBD, we have a lookahead window of 1024 blocks, and if that is exceeded, we disconnect the stalling peer.
However, if we are close to but not at the tip (<=1024 blocks), neither of these mechanisms apply. We can't do compact blocks yet, and the stalling mechanism doesn't work because the 1024 window cannot be exceeded.

As a result, we have to resort to `BLOCK_DOWNLOAD_TIMEOUT_BASE` which only disconnects a peer after 10 minutes (plus 5 minutes more for each additional peers we currently have blocks in flight). This is too long in my opinion, especially since peers get assigned up to 16 blocks (`MAX_BLOCKS_IN_TRANSIT_PER_PEER`) and could repeat this process to stall us even longer if they send us a block after 10 minutes.
This issue was observed in #29281 and https://github.com/bitcoin/bitcoin/issues/12291#issuecomment-1898517930 with broken peers that didn't send us blocks.

**Proposed solution:**
If we are 1024 or less blocks away from the tip and haven't requested or received a block from any peer for 30 seconds, add another peer to download the critical block from that would help us advance our tip. Add up to two additional peers this way.

**Other thoughts**
* I also considered the alternative of extending the existing stalling mechanism that disconnects instead of introducing parallel downloads. This could be potentially less wasteful, but we might be over-eager to disconnect peers when really close to the tip, plus this might lead to cycling through lots of peers in extreme situations where we have a very slow internet connection.
* The chosen timeout of 30 seconds could lead to inefficiencies / bandwidth waste when we have a really slow internet connection. Maybe it could make sense to track the last successful download times from existing peers and use a dynamic timeout according to that statistics, instead of setting it to a fixed value. 
* I will leave this PR in draft until I have tested it a bit more in the wild.

Fixes #29281

mzumsande committed:
p2p: Add additional peers to download from when close to the tip

If we are 1024 or less blocks away from the tip and haven't requested or received
a block from any peer for 30 seconds, add another peer to download the critical
block from. Add up to two additional peers this way.

Also adds test for the new behavior (co-authored by Greg Sanders)

Co-authored-by: Greg Sanders <gsanders87@gmail.com>

mzumsande committed:
test: make p2p_ibd_stalling.py more modular

This is in preparation to add more subtests.
Also adjust waiting condition to replace the
total_bytes_recv_for_blocks magical number hack.

mzumsande committed:
test: add sub-test for near-tip stalling

nanlour commented:
>  Maybe it could make sense to track the last successful download times from existing peers and use a dynamic timeout according to that statistics, instead of setting it to a fixed value.

Is it possible for other peers to make me believe I have a slow internet connection?

mzumsande commented:
> Is it possible for other peers to make me believe I have a slow internet connection?

Sorry, I missed this question. For single peers yes, but the idea is that the data would be taken from multiple outbound peers - it is hard for an attacker to control multiple of these, see [eclipse attacks](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Addrman-and-eclipse-attacks).

Will soon come back to this PR and take it out of draft.

mzumsande commented:
Cleaned up a bit, rebased and marked as ready for review!

glozow commented:
concept ACK

naumenkogs commented:
Concept ACK

mzumsande commented:
[824c125 ](https://github.com/bitcoin/bitcoin/commit/824c1254ed4ae248ff44ca536e82c1238b86b158)to [1bcfcf2](https://github.com/bitcoin/bitcoin/commit/1bcfcf27c134c6b78a53447bc91a78001c6181f9): Rebased

mzumsande commented:
> The CI failed, but the logs are missing :( I'll re-run 10 times.

Thanks, I can also reproduce it locally, so no need to do more CI runs!

instagibbs commented:
IIUC, this could also fire at tip if there are no hb peers and the first peer requested stalls out for 30s? Once parallel compact blocks is warmed up, the slots will likely be taken by those efforts.

If so, I wonder if this could allow us to f.e., not require a parallel compact slot be taken by an outbound, if we know we will try an outbound connection 30s later to defeat stalling behavior?

Having tests for the interplay would be interesting too.

mzumsande commented:
> IIUC(? could be off-base), this could also fire at tip if there are no hb peers and the first peer requested stalls out for 30s? Once parallel compact blocks is warmed up, the slots will likely be taken by those efforts.

Yes, I agree that it should also happen at the tip, both in the cases of low-bandwidth compact relay and in the case of no compact block relay - which seems like a good thing in most cases?! However, one possible downside might be that if you run with an extremely slow connection for which blocks typically take longer than 30s to download, it could be wasteful to add more peers.

> If so, I wonder if this could allow us to f.e., not require a parallel compact slot be taken by an outbound, if we know we will try an outbound connection 30s later to defeat stalling behavior?

I'd suspect we probably wouldn't want that, given that the additional peers introduced here are also likely to be inbound peers (if full), so attacker controlling many of our inbound slots could stall us effectively.

> Having tests for the interplay would be interesting too.

I will work on adding tests for this situation. I might also push some non-merge commits over the next days to hopefully find out why the test currently fails intermittently (still can't reproduce locally).

instagibbs commented on src/net_processing.cpp:
```diff
@@ -1623,6 +1629,29 @@ void PeerManagerImpl::FindNextBlocks(std::vector<const CBlockIndex*>& vBlocks, c
+                        GetTime<std::chrono::microseconds>() > m_last_block_requested.load() + BLOCK_NEARTIP_TIMEOUT_MAX &&
+                        GetTime<std::chrono::microseconds>() > m_last_block_received.load() + BLOCK_NEARTIP_TIMEOUT_MAX &&
+                        !already_requested_from_peer &&
+                        mapBlocksInFlight.count(pindex->GetBlockHash()) <= 2) {
```
`MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK` seems like a variable we could rename to be more general and use here, especially if both mechanisms could fire

rebroad commented:
Rather than the crude 10 minute timeout - why not actually check if the block is being downloaded from the chosen peer - i.e. work out the throughput and make a decision based on this (compared to typical throughput from other peers)?

mzumsande commented:
> I'm still not sure why it failed, tried to bump a mocktime with the last push, but poking in the dark since it doesn't fail locally for me...

Hopefullly fixed now, I believe that there was a missing sync in the added test.

> Rather than the crude 10 minute timeout - why not actually check if the block is being downloaded from the chosen peer - i.e. work out the throughput and make a decision based on this (compared to typical throughput from other peers)?

I'm not sure what exactly you mean with throughput here - do you keeping track of historical block download times, and use this as input for an algorithm for detecting stalling? If so, I'm thinking about that option (and how to exactly design an algorithm using it), and I think it could be applied to all kinds of stalling, also when deep in IBD - but it also has some downsides (more complexity, more data to keep track of, have to be careful it's not easily gameable). So I'm thinking about it but not completely sure if that is the best way to go.

In any case the current approach is more in line with the current IBD stalling detection (fixed timeouts).

mzumsande committed:
test: Add functional test for stalling at tip

When at the tip, the near-tip stalling still applies, but
it will interact with compact block requests. Add tests for this.

instagibbs reviewed (commented):
I took a bit of time to work on test coverage, with a rough branch here: https://github.com/instagibbs/bitcoin/commits/mzumsande_202403_near_tip_stalling/

last commit needs some clean up, but demonstrates some of the interplay of parallel blocks both compact and non-compact

mzumsande commented:
> I took a bit of time to work on test coverage, with a rough branch here: https://github.com/instagibbs/bitcoin/commits/mzumsande_202403_near_tip_stalling/
> 
> last commit needs some clean up, but demonstrates some of the interplay of parallel blocks both compact and non-compact

Thanks! 
I also worked on this yesterday/today and wrote a similar test, but added that as an `at_tip_stalling` subtest to `p2p_ibd_stalling.py` - while your test covering the logic in a similar way is in `p2p_compactblocks.py`. In which location do you think that it fits better (or even both?)?

Your first commit looks great, will definitely take that (but haven't yet with the latest push).

One more thing: I added the condition `m_last_block_requested.load() > 0s` so that the near-tip stalling logic can't hit right after startup when we never requested a block from anyone.

instagibbs commented:
>  while your test covering the logic in a similar way is in p2p_compactblocks.py. In which location do you think that it fits better (or even both?)?

I think your location is also fine, we're not testing that it works, just that it's being requested pretty much.

instagibbs commented:
Just for brainstorming, also made a modified branch that reduces parallel compact block downloads to a count of 2, which would allow one or more backoff blocks at tip at a less aggressive pace: https://github.com/instagibbs/bitcoin/commits/mzumsande_202403_near_tip_stalling_2/

mzumsande commented:
[973cd01 ](https://github.com/bitcoin/bitcoin/commit/973cd0174ae44ff9cbc3b41912323f71f8eafca4)to [5e1ff82](https://github.com/bitcoin/bitcoin/commit/5e1ff82251283c07274c62e1b65d5aff9be80212): 
- added more detailed coverage by @instagibbs (thanks!)
- provided missing block and check that sync finishes in the end in `near_tip_stalling` subtest
- remove unused `total_bytes_recv_for_blocks` in first commit

> Just for brainstorming, also made a modified branch that reduces parallel compact block downloads to a count of 2, which would allow one or more backoff blocks at tip at a less aggressive pace:

That sounds reasonable. Is it a problem that on slow connections or new connections with unfilled mempools the parallel compact blocks download (which is triggered immediately instead of having a timeout) would hit more often than necessary?

instagibbs commented:
> Is it a problem that on slow connections or new connections with unfilled mempools the parallel compact blocks download (which is triggered immediately instead of having a timeout) would hit more often than necessary?

I think that's already the case, a saving grace is that since we only trigger parallel cb with high-bandwidth cb peers it won't happen until you've seen a few blocks at tip at least.

rebroad commented:
> > I'm still not sure why it failed, tried to bump a mocktime with the last push, but poking in the dark since it doesn't fail locally for me...
> 
> Hopefullly fixed now, I believe that there was a missing sync in the added test.
> 
> > Rather than the crude 10 minute timeout - why not actually check if the block is being downloaded from the chosen peer - i.e. work out the throughput and make a decision based on this (compared to typical throughput from other peers)?
> 
> I'm not sure what exactly you mean with throughput here - do you keeping track of historical block download times, and use this as input for an algorithm for detecting stalling? If so, I'm thinking about that option (and how to exactly design an algorithm using it), and I think it could be applied to all kinds of stalling, also when deep in IBD - but it also has some downsides (more complexity, more data to keep track of, have to be careful it's not easily gameable). So I'm thinking about it but not completely sure if that is the best way to go.
> 
> In any case the current approach is more in line with the current IBD stalling detection (fixed timeouts).

I mean the current download speed - averaged out over the last 30 seconds should be sufficient to make a reasonably quick decision.
